api/xds_protocol.rst

```diff
@@ -220,6 +220,61 @@ the ADS server, which will be used for all resources. The :ref:`ConfigSource
 The xDS transport Protocol
 ~~~~~~~~~~~~~~~~~~~~~~~~~~
 
+.. _xds_protocol_resource_names:
+
+Resource names
+^^^^^^^^^^^^^^
+
+The xDS protocol is designed to provide a transport mechanism for opaque resources. Every resource
+has a name. This name may be opaque in simple or legacy configurations, the control plane may
+determine it and set this value in delivered configuration, e.g. RDS resource `Foo` may be a
+dependency of LDS resource `Bar`.
+
+For more advanced use cases, xDS has the concept of resource namespaces. With namespaces, resource
+names take the form of `<namespace>/<resource name>`. In this scheme, the `resource name` remains
+opaque, while the `namespace` is expected to have the form `<authority>/<qualifier>`.
+
+The `authority` field is distinguished. It is expected to be reverse DNS of the authority originating
+the resource. The rationale for the use of reverse DNS is to allow prefix namespace subscriptions
+with the ability of more specific sub-authorities to be delegated override capability. In future
+xDS extensions, the ability to sign resources may be added, this authority field will be the basis
+for validation.
+
+The `qualifier` field is opaque and may be used to further distinguish the resource namespace by
+the control plane.
+
+Resource namespaces are intended to support a number of use cases:
+
+* Delegating authority over some particular part of the global namespace, e.g. with CDS to allow
+  server X to serve up the `com.acme.foo` hierarchy.
+
+* Partitioning resource types into different classes, e.g. for distinct tenants. E.g. LDS
+  can use the namespace to determine which subset of the universe of known listeners to
+  deliver.
+
+* Support split-horizon naming of resources. Allow `com.acme.foo.bar` to override `com.acme.foo`
+  for a given resource namespace.
+
+* Avoiding the dependence on :ref:`per-node <envoy_api_msg_Core.Node>` identifiers and metadata
+  as a means of selecting resources at a management server. The use of node data is coarse
+  grained and does not scale as the number of resource types increases.
+
+
+The resource namespace is specified in a subscription's :ref:`ConfigSource
+<envoy_v3_api_field_config.core.v3.ConfigSource.namespace>` and will then appear in
+discovery requests for the subscription. Resources delivered on this subscription are said
```

Comment:
Hmmm... It seems a little inconsistent to have namespace be a separte field in the request but have it be part of the resource name in the response.  I'd prefer to pick one of these two options:

1. Have the namespace be part of the resource name in both the request and the response.
2. Have a separate field for namespace in both the request and the response.

I think that in the UDPA draft, we were planning to go with option 1.  But I can think of reasons why option 2 might be better -- e.g., the fact that in the gRPC server, we might want to get the namespace and the resource name from different sources.

Reply:
+1 I commented on this elsewhere. I have a strong preference for (2) as  matter of clarity.

Reply:
I just thought of another problem with option 2.  If we keep this a separate field in `DiscoveryRequest`, then it will be impossible for a client to subscribe to resources in different namespaces on the same xDS stream.  We would probably need to change `DiscoveryRequest` to contain something like this:

```
message ResourceMatcher {
  string namespace = 1;
  repeated string resource_names = 2;
}
repeated ResourceMatcher resource_matcher = ...;
```

This might be an argument in favor of option 1.

Reply:
FWIW, I'm working on a design doc to thrash some of this out in a clearer presentation, but I'd point out that some of the weirdness above was because I was trying to retrofit on top of the existing resource naming scheme which only accommodates a simple string (implicit inside the resource) in SoTW xDS. Incremental xDS has a first-class `Resource` object that is much more flexible. We'll need to decide how disruptive we want to be on SoTW to get the features we want.